Do less copying / zipping for Helix

dougbu opened the pull request:
- use .dotnet/ folders instead of App.Ref/App.Runtime packages for dotnet-cli layouts
  - no need to unzip files into another layout
  - fail fast or emit message when layouts don't exist
- set `$(TestDependsOnAspNetPackages)` to `false` by default
  - no need to copy all packages into publish/ folders of most test projects
    - upload very few times per queue (2 today, maybe 4 in the future), not for _every_ work item
    - reduce the size of most work item payloads, saving bandwidth and time
    - reduction should not cause problems, even around major version and TFM changes (see .dotnet/ folders)
  - add `$(TestDependsOnAspNetAppPackages)` for the App.UnitTests case; just 2 packages needed there
- get App.UnitTests working in local builds
  - grab RuntimeList.xml file from App.Runtime's obj/ folder
  - find needed packages whether running locally or on Helix agents; should always exist

nits:
- remove test infrastructure we don't need anymore
  - use one property for shared Fx and targeting pack versions; always the same
  - remove `$env:ASPNET_RUNTIME_PATH`
- rename silly ...ListListsContains... tests
- remove extra `$(TestDependsOnAspNetXyz)` settings to their new default values
- add more comments

dougbu committed:
Do less copying / zipping for Helix
- use .dotnet/ folders instead of App.Ref/App.Runtime packages for dotnet-cli layouts
  - no need to unzip files into another layout
  - fail fast or emit message when layouts don't exist
- set `$(TestDependsOnAspNetPackages)` to `false` by default
  - no need to copy all packages into publish/ folders of most test projects
    - upload very few times per queue (2 today, maybe 4 in the future), not for _every_ work item
    - reduce the size of most work item payloads, saving bandwidth and time
    - reduction should not cause problems, even around major version and TFM changes (see .dotnet/ folders)
  - add `$(TestDependsOnAspNetAppPackages)` for the App.UnitTests case; just 2 packages needed there
- get App.UnitTests working in local builds
  - grab RuntimeList.xml file from App.Runtime's obj/ folder
  - find needed packages whether running locally or on Helix agents; should always exist

nits:
- remove test infrastructure we don't need anymore
  - use one property for shared Fx and targeting pack versions; always the same
  - remove `$env:ASPNET_RUNTIME_PATH`
- rename silly ...ListListsContains... tests
- remove extra `$(TestDependsOnAspNetXyz)` settings to their new default values
- add more comments

dougbu commented:
/fyi

- Not currently separating Publish and doing that in outer builds. The separation would add needless complexity
unless up-to-date checks are expensive for inner (per-queue) builds after the first one (despite matching global
properties).
- Not moving package uploads into a Helix correlation payload because the requirements are per-work item (test project).

HaoK commented on src/Framework/test/Microsoft.AspNetCore.App.UnitTests.csproj:
```diff
@@ -1,14 +1,17 @@
     <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
     <RootNamespace>Microsoft.AspNetCore</RootNamespace>
     <VerifyAncmBinary Condition="'$(TargetOsName)' == 'win' AND '$(TargetArchitecture)' != 'arm'">true</VerifyAncmBinary>
-    <TestDependsOnAspNetPackages>true</TestDependsOnAspNetPackages>
```
One thing to consider is to just stop running these special app tests on helix if these are now the only ones that require all this custom packaging, and just run these on the build agents once (say the windows test job).  Might save us a bunch of grief in the long term maintaining this stuff if we no longer need it for any of the other normal tests

dougbu replied:
> One thing to consider is to just stop running these special app tests on helix if these are now the only ones that require all this custom packaging, and just run these on the build agents once (say the windows test job). Might save us a bunch of grief in the long term maintaining this stuff if we no longer need it for any of the other normal tests

There's now very little infrastructure for this and any move away from Helix is a regression. In addition, the project template tests need more stuff than these tests.

HaoK reviewed (approved):
Looks great to me, less magic and less setup that we have to do for almost all the tests!

dougbu commented on eng/helix/helix.proj:
```diff
@@ -79,9 +78,12 @@
 
-  <Target Name="IncludeAspNetRuntime" BeforeTargets="Gather"
-    Condition="'$(DoNotRequireSharedFxHelix)' != 'true' OR
-    EXISTS('$(RepoRoot)artifacts\packages\$(Configuration)\Shipping\Microsoft.AspNetCore.App.Runtime.$(TargetRuntimeIdentifier).$(SharedFxVersion).nupkg')">
```
@BrennanConroy I realize this line didn't work despite my insistence on including it in your `$(DoNotRequireSharedFxHelix)` work. `$(SharedFxVersion)` was never set which meant RunHelix.ps1 couldn't successfully run App.Unitest or the project template tests ☹️

dougbu commented on eng/helix/helix.proj:
```diff
@@ -90,32 +92,39 @@
-    <ItemGroup>
-      <HelixCorrelationPayload Include="$(OutputPath)\SharedFx.Layout.zip" Destination="$(DotNetCliDestination)" />
+    <!-- After complaining about missing App.Ref layout, do nothing else in this target.  -->
+    <Warning Text="Targeting path location '$(_AppRefPath)' does not exist."
```
@BrennanConroy and others: Any thoughts on whether this should instead be a visible `<Message Importance="High" ... />` instead❔

BrennanConroy replied:
Should probably change the message to mention that the test run wont be using the local build of our shared fx and will use XYZ instead

dougbu replied:
This isn't a .NET SDK project and would need to read artifacts/bin/GenerateFiles/* to figure out the default when 7.0.0-ci isn't available. How about

```suggestion
    <Warning Text="Targeting path location '$(_AppRefPath)' does not exist and tests will use the SDK's bundled runtimes and targeting pack."
```

or (same wording but answering my own question to avoid the warnings-as-error elevation)

```suggestion
    <Message Importance="High" Text="Targeting path location '$(_AppRefPath)' does not exist and tests will use the SDK's bundled runtimes and targeting pack."
```

Could also add that tests which absolutely require the just-built runtime (most obviously the project template tests) are doomed to fail.

dougbu replied:
Thanks for the thumbs up @BrennanConroy. Do you prefer `<Warning ... />` or `<Message Importance="High" ... />`❔

BrennanConroy replied:
If this happens on the CI do we want to know about it? Then warning. Otherwise, whichever is fine.

dougbu replied:
> If this happens on the CI do we want to know about it?

Yes because it dooms a subset of the tests to fail and it's not worth running the rest. I'm going to use slightly stronger language in the `<Error/>` two lines down.

To clarify: The `<Error/>` happens on the CI or if you add `/p:DoNotRequireSharedFxHelix=false` to a RunHelix.ps1 command.

dougbu committed:
!fixup! Normalize paths for Helix
- use `$(ArtifactsShippingPackagesDir)` in Helix.targets
- use `[MSBuild]::NormalizeDirectory(...)` and `$([System.IO.Path]::Combine(...)` in helix.proj

dougbu committed:
Review comments: Reword warning and error
- explain a bit more about missing .dotnet/ layouts
